### ISSUE TITLE
Hashing of unhashable objects is not detected

### ISSUE THREAD

**Issue by gsakkis** [python/mypy#20230]
**Bug Report**

False negatives for hashing unhashable objects

**To Reproduce**

```python
class Unhashable1:
    def __eq__(self, other: object) -> bool:
        return False


class Unhashable2:
    __hash__ = None  


for obj in Unhashable1(), Unhashable2():
    try:
        hash(obj)
    except TypeError:
        pass
    else:
        raise AssertionError(f"{obj} should be unhashable")

    try:
        obj in set()
    except TypeError:
        pass
    else:
        raise AssertionError(f"{obj} should be unhashable")

    try:
        obj in {}
    except TypeError:
        pass
    else:
        raise AssertionError(f"{obj} should be unhashable")

    try:
        s = {obj}  
    except TypeError:
        pass
    else:
        raise AssertionError(f"{obj} should be unhashable")

    try:
        d = {obj: "value"}
    except TypeError:
        pass
    else:
        raise AssertionError(f"{obj} should be unhashable")
```

**Expected Behavior**

5 errors to be reported

**Actual Behavior**

No errors reported (other than the [`__hash__ = None`](https://github.com/python/mypy/issues/4266) false positive)

**Your Environment**

- Mypy version used: 1.18.2
- Python version used: 3.12
